src/vos/vos_io.c

```diff
@@ -78,6 +78,203 @@ struct vos_io_context {
+	int		 de_ref;
+};
+
+static inline struct dedup_entry *
+dedup_rlink2entry(d_list_t *rlink)
+{
+	return container_of(rlink, struct dedup_entry, de_link);
+}
+
+static bool
+dedup_key_cmp(struct d_hash_table *htable, d_list_t *rlink,
+	      const void *key, unsigned int csum_len)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+	struct dcs_csum_info	*csum = (struct dcs_csum_info *)key;
+
+	D_ASSERT(entry->de_csum_len != 0);
+	D_ASSERT(csum_len != 0);
+
+	if (entry->de_csum_len != csum_len)
+		return false;
+
+	D_ASSERT(csum->cs_csum != NULL);
+	D_ASSERT(entry->de_csum_buf != NULL);
+
+	return memcmp(entry->de_csum_buf, csum->cs_csum, csum_len) == 0;
+}
+
+static uint32_t
+dedup_key_hash(struct d_hash_table *htable, const void *key,
+	       unsigned int csum_len)
+{
+	struct dcs_csum_info	*csum = (struct dcs_csum_info *)key;
+
+	D_ASSERT(csum_len != 0);
+	D_ASSERT(csum->cs_csum != NULL);
+
+	return d_hash_string_u32((const char *)csum->cs_csum, csum_len);
+}
+
+static void
+dedup_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+
+	entry->de_ref++;
+}
+
+static bool
+dedup_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+
+	D_ASSERT(entry->de_ref > 0);
+	entry->de_ref--;
+
+	return entry->de_ref == 0;
+}
+
+static void
+dedup_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dedup_entry	*entry = dedup_rlink2entry(rlink);
+
+	D_ASSERT(entry->de_ref == 0);
+	D_ASSERT(entry->de_csum_buf != NULL);
+
+	D_FREE(entry->de_csum_buf);
+	D_FREE(entry);
+}
+
+static d_hash_table_ops_t dedup_hash_ops = {
+	.hop_key_cmp	= dedup_key_cmp,
+	.hop_key_hash	= dedup_key_hash,
+	.hop_rec_addref	= dedup_rec_addref,
+	.hop_rec_decref	= dedup_rec_decref,
+	.hop_rec_free	= dedup_rec_free,
+};
+
+int
+vos_dedup_init(struct vos_pool *pool)
+{
+	int	rc;
+
+	rc = d_hash_table_create(D_HASH_FT_NOLOCK, 13, /* 8k buckets */
+				 NULL, &dedup_hash_ops,
+				 &pool->vp_dedup_hash);
+
+	if (rc)
+		D_ERROR(DF_UUID": Init dedup hash failed. "DF_RC".\n",
+			DP_UUID(pool->vp_id), DP_RC(rc));
+	return rc;
+}
+
+void
+vos_dedup_fini(struct vos_pool *pool)
+{
+	if (pool->vp_dedup_hash) {
+		d_hash_table_destroy(pool->vp_dedup_hash, true);
+		pool->vp_dedup_hash = NULL;
+	}
+}
+
+void
+vos_dedup_invalidate(struct vos_pool *pool)
+{
+	vos_dedup_fini(pool);
+	vos_dedup_init(pool);
+}
+
+static bool
+vos_dedup_lookup(struct vos_pool *pool, struct dcs_csum_info *csum,
+		 daos_size_t csum_len, struct bio_iov *biov)
+{
+	struct dedup_entry	*entry;
+	d_list_t		*rlink;
+
+	if (!ci_is_valid(csum))
+		return false;
+
+	rlink = d_hash_rec_find(pool->vp_dedup_hash, csum, csum_len);
+	if (rlink == NULL)
+		return false;
+
+	entry = dedup_rlink2entry(rlink);
+	*biov = entry->de_biov;
+
+	D_ASSERT(entry->de_ref > 1);
+	D_ASSERT(biov->bi_addr.ba_dedup);
+	D_ASSERT(biov->bi_buf == NULL);
+
+	d_hash_rec_decref(pool->vp_dedup_hash, rlink);
```

Comment:
should this not be an addref

Reply:
nevermind, it's copying the bio data.   We will need a persistent reference count for this to work in production but this is just for quick lookup